Correctly dispatch the printing of custom vs built-in functions

normalhuman opened the pull request:
Undefined functions (whether applied or not) are not longer sent to custom `_print_foo` methods designed for printing built-in functions. Additionally, correct printing for beta function is added for both LaTeX and pretty print (this function incorrectly printed as lowercase beta, in all printers). The logic for LaTeX printing of un-applied built-in functions like gamma or DiracDelta is moved from assorted classes to LaTeX printer. The same logic is now applied to pretty printer as well. Fixes #2832.

#### Minor changes

#### Centralization of printing rules

Although it's been decided long ago that the LaTeX printing logic should be in LaTeX printer, some classes such as gamma function retained it. (As a result, the codebase had two functions responsible solely for LaTeX printing gamma). Now the methods `_latex` and `_latex_no_arg` are removed from the special functions module and their content is expressed in latex.py

#### All printers

The printer driver in printer.py uses `type(expr).__mro__` to locate an appropriate custom printer, trying the classes in order. Since for `Function('gamma')` the class 'gamma' precedes "UndefinedFunction" in the MRO list, this resulted in an incorrect dispatch responsible for #2832. Change: now, if the MRO list contains AppliedUndef or UndefinedFunction, their subclasses are ignored for printing purposes.

#### LaTeX printing

- The method `_print_Function` of LaTeX printer no longer sends undefined functions to custom print methods
- The method  `_print_FunctionClass` (for unapplied functions) has a dictionary of rules for Kronecker Delta, gamma, lowergamma, beta, DiracDelta, and Chi (the hyperbolic cosine integral), which were previously scattered among those classes.
- Added methods `_print_beta` and `_print_Chi` for beta function and hyperbolic cosine integral (applied versions). The correct symbol for [beta function](https://en.wikipedia.org/wiki/Beta_function) is uppercase Greek letter Beta, which unfortunately is absent in LaTeX. `\operatorname{B}` is used as an approximation to it.

#### Pretty Printing

- Added method `_print_FunctionClass` paralleling the LaTeX treatment of certain unapplied functions. Previously there was no such logic: for example, `pprint(gamma)` printed as "gamma" not as Γ.
- Added an optional keyword argument `func_name` to `_print_Function` method, so that some `_print_foo` methods could reduced to simply passing a suitable Unicode/non-Unicode name to this function. Besides shorter code, this improved the consistency of rendering: previously, the exponent 2 in `pprint(gamma(x)**2)` was positioned differently in Unicode and non-Unicode versions.
- Corrected the names of beta and gamma functions. Previously, gamma was printed as "gamma" instead of "Gamma" in non-Unicode mode. And beta was printed as "beta" instead of Greek letter Beta (Unicode) or Latin letter B (non-Unicode). Also, `_print_gamma` is simplified because it no longer receives incorrectly dispatched undefined functions.

#### Tests

New tests for beta and gamma are added. Some tests in crv_types.py were corrected because they had the incorrect notation for beta and gamma functions baked in.

normalhuman committed:
Correctly dispatch the printing of custom vs built-in functions

Undefined functions (whether applied or not) are not longer sent
to custom _print_foo methods designed for printing built-in functions.
Additionally, correct printing for beta function is added for both
LaTeX and pretty print (this function incorrectly printed as
lowercase beta, in all printers). The logic for LaTeX printing of
un-applied built-in functions like gamma or DiracDelta is moved
from assorted classes to LaTeX printer. The same logic is now applied
to pretty printer as well. Fixes #2832 and miscellaneous minor bugs.

normalhuman committed:
Correct off by one error in slicing the MRO list

Adding 1 to the index found was not needed.

ylemkimon commented on sympy/printing/latex.py:
```diff
@@ -767,10 +768,21 @@ def _print_UndefinedFunction(self, expr):
+                         lowergamma: r'\gamma',
+                         beta: r'\operatorname{B}',
+                         DiracDelta: r'\delta',
+                         Chi: r'\operatorname{Chi}'}
```
I think these should be centralized. Could you move them to the top and use them in the respective function printing methods, or use `_print_FunctionClass` for printing function's name in their printing methods?

ylemkimon replied:
Or I think we can move special functions handling except `KroneckerDelta` and `DiracDelta` to `_print_Function`, as their methods are identical and probably other special functions we might add will be same too.

normalhuman replied:
I like the idea of centralizing things even further, i.e., removing the various `_print_gamma`, `_print_lowergamma` etc and just adding a dictionary of special cases to `_print_Function`.  Most of them don't do anything special apart from custom name. But this feels like a separate project to me; this PR already touches many things at once.

asmeurer replied:
I think @ylemkimon just means to move this dictionary up to the class level, so it is easier for subclasses to override (otherwise they'd have to redefine this entire function).

normalhuman replied:
Circular imports seem to prevent me from doing so.  Hopefully not much of an issue, because this function does not have much besides those imports and that dictionary.

asmeurer replied:
You can work around that by making it a @property method. It won't be a class attribute in that case, but I don't think that's a big deal.

asmeurer commented:
I agree that the special cases logic should be moved out of the functions.

asmeurer commented on sympy/printing/latex.py:
```diff
@@ -767,10 +768,21 @@ def _print_UndefinedFunction(self, expr):
+                         beta: r'\operatorname{B}',
+                         DiracDelta: r'\delta',
+                         Chi: r'\operatorname{Chi}'}
+        if expr in special_cases:
```
Should probably do a subclass check. Although if the subclass name is different it should print the actual class name.

normalhuman replied:
I'm weak in subclassing; do you have an example in mind of the kind of things to check for? This is within `_print_FunctionClass` so if the printer got here, that means `type(expr).__mro__` did not have any subclass with a custom _print_ implemented. If someone subclasses, say, DiracDelta and doesn't add a _print_ for their new class, it will print as `r'\delta'.

normalhuman replied:
I now see what you mean. Added a commit replacing direct comparison by `issubclass` (both in latex and pretty) and added tests.

asmeurer replied:
I mean like

```py
class beta(sympy.beta):
    ...
```

should still print as `\beta`, but

```py
class mybeta(sympy.beta):
    ...
```

shouldn't (because the name changed to `mybeta`).

normalhuman replied:
I see. That's easy to make work for unapplied (FunctionClass) but to make `mybeta(x, y)` behave in the same way I'll need to remove all the _print_beta, _print_gamma, etc moving them to some special_function dictionary used by _print_Function. Otherwise, the driver following `(mybeta, beta, Function,..)` hierarchy will dispatch mybeta to _print_beta, not to _print_Function.

But it still won't work well because customizations are not just name changes; say, KroneckerDelta has its arguments placed as subscripts. Should this argument-placement behavior be inherited by subclasses that have a different name? Unclear to me. 

Considering all this, I think I prefer the original approach without subclass check: no special printing rules pass to subclasses.

normalhuman replied:
The current behavior of functions is that notation passes to subclasses, regardless of how they are named.
```
class myg(gamma):
  pass
latex(myg(x))    r'\Gamma{\left(x \right)}'
```
To change the printed name, one would have to add `_print_myg` method to each printer. To my understanding, this is the trade-off that was [consciously made](http://docs.sympy.org/dev/modules/printing.html):

> works great for built-in sympy objects, but not that good for user defined classes where it is inconvenient to patch the printers.

normalhuman replied:
I think if we want to change this notation inheritance, the most practical way is to truncate the MRO list in printer.py, similar to how it was done for undefined functions. The rule would be: when MRO contains `Function`, ignore the subclasses of Function except for the smallest one. So instead of `(mybeta, beta, Function,..)` the list is `(mybeta, Function,..)` and since there is no _print_mybeta, it gets dispatched to _print_Function which looks at the function name. Does this sound reasonable?

asmeurer commented on sympy/printing/pretty/pretty.py:
```diff
@@ -1267,38 +1290,21 @@ def _print_SingularityFunction(self, e):
-            return pform
-        else:
-            return self._print_Function(e)
+        func_name = greek_unicode['Gamma'] if self._use_unicode else 'Gamma'
```
Don't know how I feel about the ASCII pretty printer changing to Gamma.

normalhuman replied:
My thinking was: we want to type Γ but don't have Unicode, what do we do? I think I would type "Gamma" in that situation, not "gamma". Pretty printing is about math notation, not the names of Python objects.

normalhuman committed:
Notation for FunctionClass special cases passes to subclasses

Changed the direction comparison of FunctionClass to issubclass check,
so that special notation is inherited by subclasses. Tests added.

normalhuman committed:
Remove redundant isinstance condition on expr.func

isinstance(expr.func, UndefinedFunction) is enough to detect we have
an undefined function

normalhuman committed:
Remove redundant isinstance condition on expr.func

isinstance(expr.func, UndefinedFunction) is enough to detect we have
an undefined function

ylemkimon commented on t.py:
```diff
@@ -0,0 +1,5 @@
+x, g, K, m, t = sym.symbols('x, g, K, m, t', positive=True, real=True)
+dxx = sym.Eq(x(t).diff(t, 2), g - K/m*x(t).diff(t))
+ics = {x(0): 0, x(t).diff(t).subs(t, 0): 0}
+print(sym.dsolve(dxx, ics=ics))
```
This file seems to be committed by mistake.

normalhuman replied:
Thanks for noticiing, I'll try to get rid of it.

normalhuman committed:
Merge branch '2832_latex_undefined_function' of github.com:normalhuman/sympy into 2832_latex_undefined_function

normalhuman committed:
remove a mistakenly committed file

normalhuman committed:
Make the dictionary of special function classes a property of printer

Moved the dictionary of function classes that have special notation
to a @property, in both LaTeX and pretty printers.

normalhuman committed:
Do not pass function notation to subclasses of a different name

If someone subclasses a function, say, gamma, and calls it mygamma
instead, it will no longer print as \gamma in LaTeX (similarly for
pretty printer).

normalhuman committed:
Also test notation inheritance for a subclass with the same name

Added a test to make sure that subclassing without renaming the
subclass preserves custom printing for that class. An already
present test ensures that when the name is different, notation is
not inherited.

asmeurer commented:
I don't feel strongly either way about the gamma thing. My main comments here were to properly handle subclasses and to factor out the variable to the class level, which have both been addressed.

moorepants commented:
Seeing some performance regressions from this PR, see #14671.
